hnn_core/extracellular.py

```diff
@@ -0,0 +1,601 @@
+        # b: position vector of electrode with respect to section end (1)
+        # H: parallel distance from section end to electrode
+        # R: radial distance from section end to electrode
+        # L: parallel distance from section start to electrode
+        # Note that there are three distinct regimes to this approximation,
+        # depending on the electrode position along the section axis.
+        phi = np.zeros(len(seg_ctr))
+        for idx, (ctr, line_len) in enumerate(zip(seg_ctr, line_lens)):
+            start = ctr - line_len * sec_vec / norm(sec_vec)
+            end = ctr + line_len * sec_vec / norm(sec_vec)
+            a = end - start
+            norm_a = norm(a)
+            b = electrode_pos - end
+            # projection: H = a.cos(theta) = a.dot(b) / |a|
+            H = np.dot(b, a) / norm_a  # NB can be negative
+            L = H + norm_a
+            R2 = np.dot(b, b) - H ** 2  # NB squares
+
+            # To avoid very large values when electrode is placed (anywhere) on
+            # the section axis, enforce minimal perpendicular distance
+            R2 = np.maximum(R2, min_distance ** 2)
+
+            if L < 0 and H < 0:  # electrode is "behind" line segment
+                num = np.sqrt(H ** 2 + R2) - H  # == norm(b) - H
+                denom = np.sqrt(L ** 2 + R2) - L
+            elif L > 0 and H < 0:  # electrode is "on top of" line segment
+                num = (np.sqrt(H ** 2 + R2) - H) * (L + np.sqrt(L ** 2 + R2))
+                denom = R2
+            else:  # electrode is "ahead of" line segment
+                num = np.sqrt(L ** 2 + R2) + L
+                denom = np.sqrt(H ** 2 + R2) + H  # == norm(b) + H
+
+            phi[idx] = np.log(num / denom) / norm_a
+
+    # [dis]: um; [sigma]: S / m
+    # [phi / sigma] = [1/dis] / [sigma] = 1 / [dis] x [sigma]
+    # [dis] x [sigma] = um x (S / m) = 1e-6 S
+    # transmembrane current returned by _ref_i_membrane_ is in [nA]
+    # ==> 1e-9 A x (1 / 1e-6 S) = 1e-3 V = mV
+    # ===> multiply by 1e3 to get uV
+    return 1000.0 * phi / (4.0 * np.pi * sigma)
+
+
+class ExtracellularArray:
+    """Class for recording extracellular potential fields with electrode array
+
+    Note that to add an electrode array to a simulation, you should use the
+    :meth:`hnn_core.Network.add_electrode_array`-method. After simulation,
+    the network will contain a dictionary of `ExtracellularArray`-objects
+    in ``net.rec_array`` (each array must be added with a unique name). An
+    `ExtracellularArray` contains the voltages at each electrode contact,
+    along with the time points at which the voltages were sampled.
+
+    Parameters
+    ----------
+    positions : tuple | list of tuple
+        The (x, y, z) coordinates (in um) of the extracellular electrodes.
+    sigma : float
+        Extracellular conductivity, in S/m, of the assumed infinite,
+        homogeneous volume conductor that the cell and electrode are in.
```

Comment:
Is there a reason why we're not just calling this `g_extracell`, `extracell_conduc`, or something that is a little more explicit? Since `sigma` is already used to denote time stdev elsewhere in the model, we may want to avoid using it here.

Reply:
how about `conductivity` to be explicit and to [mirror](https://mne.tools/stable/generated/mne.make_bem_model.html#mne.make_bem_model) MNE API

Reply:
I like `conductivity`. The `gxx` should be reserved for conductance, which is in different physical units (or at least can be). Anyway it's microscopic, whereas conductivity is a bulk property.

Reply:
`conductivity` sounds good!